field_audit/models.py

```diff
@@ -0,0 +1,261 @@
+            # both
+            return self._by_type_and_username(user_type, username)  # pragma: no cover  # noqa: E501
+
+
+class DefaultAuditEventManager(AuditEventManager):
+    """Default Manager for the AuditEvent model. Contains convenience methods
+    for the default auditors, which may not be desirable to subclass if
+    downstream projects wish to define custom auditor chains.
+    """
+
+    def by_system_user(self, username):
+        system_types = [USER_TYPE_TTY, USER_TYPE_PROCESS]
+        return self.filter(
+            changed_by__user_type__in=system_types,
+            changed_by__username=username,
+        )
+
+    def by_tty_user(self, username):
+        return self.by_type_and_username(USER_TYPE_TTY, username)
+
+    def by_process_user(self, username):
+        return self.by_type_and_username(USER_TYPE_PROCESS, username)
+
+    def by_request_user(self, username):
+        return self.by_type_and_username(USER_TYPE_REQUEST, username)
+
+
+def get_manager(attr_suffix, default):
+    """Returns an instantiated manager, possibly one defined in settings.
+
+    If the manager is defined in settings, it must be a subclass of
+    ``django.db.models.Manager``.
+
+    :param attr_suffix: Suffix (appended to ``FIELD_AUDIT_``) of settings
+        attribute for the manager class path.
+    :param default: Manager class. Returned if attribute isn't defined.
+    """
+    settings_attr = f"FIELD_AUDIT_{attr_suffix}"
+    try:
+        class_path = getattr(settings, settings_attr)
+    except AttributeError:
+        return default()
+    desc = f"{settings_attr!r} value"
+    return class_import_helper(class_path, desc, models.Manager)()
+
+
+def get_date():
+    """Returns the current UTC date/time.
+
+    This is the "getter" for default values of the ``AuditEvent.event_date``
+    field.
+    """
+    return datetime.utcnow()
+
+
+class AuditEvent(models.Model):
+    event_date = models.DateTimeField(default=get_date, db_index=True)
+    object_class_path = models.CharField(db_index=True, max_length=255)
+    object_pk = models.JSONField()
+    changed_by = models.JSONField()
+    is_create = models.BooleanField(default=False)
+    is_delete = models.BooleanField(default=False)
+    delta = models.JSONField()
+
+    objects = get_manager("AUDITEVENT_MANAGER", DefaultAuditEventManager)
+
+    class Meta:
+        constraints = [
+            models.CheckConstraint(
+                name="field_audit_auditevent_valid_create_or_delete",
+                check=~models.Q(is_create=True, is_delete=True),
+            ),
+        ]
+
+    ATTACH_INIT_VALUES_AT = "__field_audit_init_values"
+
+    @staticmethod
+    def get_field_value(instance, field_name):
+        """Returns the database value of a field on ``instance``.
+
+        :param instance: an instance of a Django model
+        :param field_name: name of a field on ``instance``
+        """
+        return instance._meta.get_field(field_name).value_from_object(instance)
```

Comment:
what does this return for foreign key fields?

Reply:
It returns the "referenced" value. That is, the value of `<column>` in the example SQL snip: `FOREIGN KEY REFERENCES <table>(<column>)`.

This is often the primary key of the "referenced" record, but foreign keys don't have to reference a primary key, and this is how you get that value regardless of what the `REFERENCES` column is.